src/lsst/rsp/utils.py

```diff
@@ -73,63 +81,99 @@ def jupyter_proxy_url(port: Optional[int]) -> str:
-    return get_pod().spec.node_name
+    pod = get_pod()
+    if pod is not None:
+        return pod.spec.node_name
+    # In Nublado v3, we push this into the environment as K8S_NODE_NAME
+    return os.environ.get("K8S_NODE_NAME", "")
 
 
 def get_digest() -> str:
     """Extract image digest from pod, if we can."""
+    pod = get_pod()
+    # Image ID looks like host/[project/]owner/repo@sha256:hash
+    if pod is None:
+        # In Nublado v3 we push the digest directly into the image spec
+        image_id = os.environ.get("JUPYTER_IMAGE_SPEC", "")
+    else:
+        try:
+            image_id = pod.status.container_statuses[0].image_id
+        except Exception:
+            image_id = ""
     try:
-        img_id = get_pod().status.container_statuses[0].image_id
-        # host/owner/repo@sha256:hash
-        return (img_id.split("@")[-1]).split(":")[-1]
+        return (image_id.split("@")[-1]).split(":")[-1]
     except Exception:
         return ""  # We will just return the empty string
 
 
 def get_access_token(
     tokenfile: Optional[Union[str, Path]] = None, log: Optional[Any] = None
-) -> Optional[str]:
-    """Determine the access token from the mounted configmap (nublado2),
-    secret (nublado1), or environment (either).  Prefer the mounted version
-    since it can be updated, while the environment variable stays at whatever
-    it was when the process was started."""
-    tok = None
+) -> str:
+    """Determine the access token from the mounted location (nublado
+    3/2) or environment (any).  Prefer the mounted version since it
+    can be updated, while the environment variable stays at whatever
+    it was when the process was started.  Return the empty string if
+    the token cannot be determined.
+    """
+    tok = ""
     if tokenfile:
         # If a path was specified, trust it.
         tok = Path(tokenfile).read_text()
         tried_path = tokenfile
     else:
-        # Try the default token paths, nublado2 first, then nublado1
-        n2_tokenfile = "/opt/lsst/software/jupyterlab/environment/ACCESS_TOKEN"
-        tried_path = n2_tokenfile
-        token_path = Path(n2_tokenfile)
-        try:
-            tok = token_path.read_text()
-        except Exception:
-            # OK, it's not mounted.  Fall back to the environment.
-            pass
-    if not tok:
-        tok = os.environ.get("ACCESS_TOKEN", None)
-    if not tok:
-        raise ValueError(
-            f"Could not find token in env:ACCESS_TOKEN nor in {tried_path}"
-        )
+        jldir = "/opt/lsst/software/jupyterlab"
+        tried_path = ""
+        # Try the default token paths: nublado3, then nublado2, then fall
+        # back to the environment.
+        tokenfiles = [
+            f"{jldir}/secrets/token",
+            f"{jldir}/environment/ACCESS_TOKEN",
+        ]
+        for tf in tokenfiles:
+            if not tried_path:
+                tried_path = tf
+            else:
+                tried_path += f", {tf}"
```

Comment:
The `tried_path` bookkeeping doesn't appear to be used by anything.

Reply:
Ah, right.  It used to be part of a raised exception, but I decided just returning the empty string if we can't get the token would be better.

Reply:
Also line 171 should have been outdented.  Fixed.